hicommonwealth/commonwealth
# New Thread Previews with recent comments, for community discussion, community overview and user dashboard pages

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/7865

## Description of Changes
- Added new thread previews in Community Discussion, Community Overview, and User Dashboard "For-You" and "Global" sections, this would display 3 most recent comments with comment author and profile details on the thread
- Updated react queries to reflect the latest comment state in recent comments preview of thread card
- Decoupled most of `<ThreadCard/>`, `<ThreadOptions/>`, `<AuthorAndPublishInfo/>`, `<CommentCard/>` from `app.activeChainId()` context, which makes them useable in global non-community context.
- Updated `/threads?bulk=true` and `/threads?active=true` to return 3x most recent comments on the thread
- Updated `PATCH /threads` API to return 3x recent comments with comment author and profile details on threads and thread reactions. This also fixed a bug with thread updates that would remove reactions from the thread once a thread gets updated. 
- Updated `/viewUserActivity` and `/viewGlobalActivity` API endpoints to return data relevant for thread preview on user dashboard sections.
- Updated most methods of Permission utility to work in the global non-community context.
- Updated thread and comment share component to work in the global non-community context + This also fixed a bug on the view threads page when sharing a thread would give undefined in the shared link
- Updated thread and comments disabled action tooltips across the whole component usage, this fixed a bug with comment actions which were giving incorrect tooltip text when actions were disabled
- Fix `app.chain.accounts` checks in the upvotes drawer which were breaking due to nonexisting user address details
- Added topic gating checks for thread cards preview on user dashboard "for you" and "global" sections and community overview page

## "How We Fixed It"
N/A

## Test Plan
This PR should be tested with 
- A dedicated site admin account
- A dedicated community admin account
- A dedicated community moderator account
- A dedicated site member account
- With and without a custom domain setup

Note: Some of the steps below may feel too basic, but they also break coz of the nature of different thread responses we manage in the frontend and those that come from API. Testing all of these is important

1. Without the custom domain setup
2. Login with a normal account
3. Visit `/dydx` or any other community
4. Create some threads in different topics
5. Add comments to those threads with your current account, and log in with another account to add some more comments
6. Now visit the `/discussions` page of that community, verify you see the new thread preview, with 3x recent comments on each thread
7. Now visit the `/overview` page of that community and verify you see the same new thread preview, with 3x recent comments on each thread but with a topic separation layout
8. Now visit the `/dashboard/for-you` and `/dashboard/global` pages of the user dashboard and verify you see the same new thread preview with 3x recent comments on each thread.
9. Now for all thread previews you saw in steps 6, 7, and 8, verify
     - Clicking on the `comment` button takes you to the thread details page, with the comment editor in focus
     - Clicking on the `subscribe` button, subscribes you to the thread
     - Clicking on the `share` button opens a popover, and the copied link and the Twitter share link match with the link of the thread details page. 
     - Clicking on the `reply` button on the recent comment, redirects you to the thread details page with that specific comment in focus. 
     - Clicking on the `share` button on a recent comment opens a popover, and the copied link and the Twitter share link match with the link of the thread details page with that specified comment in focus. Visiting that link opens the thread details page and focuses that specific comment 
     - Clicking on the thread card redirects to the thread details page
     - Clicking on the comment card redirects to the thread details page with that specific comment in focus
     - Verify the thread previews in the community overview and user dashboard sections don't have an `upvote`, `view upvotes` and `modify-actions` buttons 
     - Verify the author and user info section of the thread preview in user dashboard sections have the community details, while other thread previews don't
10. Now select a specific thread and test updates on it, verify
     - Updating thread title reflects in all previews you saw in 6, 7, and 8
     - Updating thread body reflects in all previews you saw in 6, 7, and 8
     - Updating thread spam status reflects in all previews you saw in 6, 7, and 8
     - Updating thread lock status reflects in all previews you saw in 6, 7, and 8
     - Updating thread stage status reflects in all previews you saw in 6, 7, and 8
     - Updating thread archive status reflects in all previews you saw in 6, 7, and 8
     - Updating thread pin status reflects in all previews you saw in 6, 7, and 8
     - Updating a comment in that thread reflects in all previews you saw in 6, 7, and 8
     - Updating a comment spam status in that thread reflects in all previews you saw in 6, 7, and 8
11. Test step 10 with a site admin account, a community admin account, a community moderator account, and that specific thread author and collaborator account.
12. Now add topic gating to the topic which was using the thread we selected in 10, and verify that if you don't meet the Gating threshold 
    - with a normal account, the thread actions and its comment preview actions in all previews you saw in 6, 7, and 8 are disabled with the `Topic is gated` tooltip.
    - with a site admin, or community admin, or community moderator,  or thread author, or thread collaborator account, the thread actions and its comment preview actions in all previews you saw in 6, 7, and 8 are enabled.
13. Now lock the thread we selected in 10 and verify that the thread actions and its comment preview actions in all previews you saw in 6, 7, and 8 are disabled with the `Thread is locked` tooltip.
14. Now archive the thread we selected in 10 and verify that the thread actions and its comment preview actions in all previews you saw in 6, 7, and 8 are disabled with the `Thread is archived` tooltip.
15. Now open the thread we selected in 10 in an incognito tab and verify that the thread actions and its comment preview actions in all previews you saw in 6, 7, and 8 are disabled with `Join community to ${'upvote' | 'comment' | 'reply' | 'submit'}` tooltip depending on the action.
16. Verify all thread previews you saw in 6, 7, and 8 are responsive across different screen sizes
17. Test all remaining actions of the thread and comments and verify you see consistent behavior in all thread previews you saw in 6, 7, and 8 
18. Now repeat steps 2 to 17 with a site admin account, a community admin account, a community moderator account, and a custom domain setup. Do each of these separately. When testing with a custom domain setup, also verify all the thread and comment links point to the correct URLs

The above steps should cover all the test cases, but plz test if you feel any related logic might break

## Deployment Plan
N/A

## Other Considerations
N/A
